File: apps/reading/src/api/antiekBenchRewriteModelDecisionMarketplaceCompose.ts
/**
 * Antiek-bench recursive rewrite residual over model decision-tree budget +
 * twin search + HTML-native recursive twin marketplace free pack (pure).
 *
 * Operator vision: weekly Antiek-bench learns from usage patterns (what
 * worked / didn't) and proposes sub-benchmark rewrites as the platform
 * expands — stacked on model driver selection with usage bar + projection,
 * twin intelligent search, HTML-native view, recursive twin note-taker, and
 * free-before-buy marketplace honesty.
 *
 * suite_rewritten / applied always false (proposal only).
 * live_router_authorized / secrets_stored / live_meter_read always false.
 * remote_index_queried / pdf_primary always false.
 * production_router_verdict always REJECT.
 */

import {
  proposeAntiekBenchRecursiveRewrite,
  type BenchRewriteProposal,
  type UsagePattern,
} from "./antiekBenchRecursiveRewrite";
import {
  composeModelDecisionTwinSearchHtmlNativeMarketplace,
  type ModelDecisionTwinSearchHtmlNativeMarketplaceCompose,
  type ModelDecisionTwinSearchHtmlNativeMarketplaceInput,
} from "./modelDecisionTwinSearchHtmlNativeMarketplaceCompose";

export interface AntiekBenchRewriteModelDecisionMarketplaceInput {
  /** Usage patterns that drive recursive rewrite proposals (never invented). */
  rewrite: {
    week_label: string;
    patterns: UsagePattern[];
  };
  model_decision_pack: Omit<
    ModelDecisionTwinSearchHtmlNativeMarketplaceInput,
    "operator_ack"
  >;
  operator_ack: boolean;
  /**
   * When true (default), require rewrite has ≥1 proposal AND
   * model_decision_pack.pack_ready.
   */
  require_both?: boolean;
  /**
   * When true (default), block pack if rewrite would apply production changes
   * (applied must remain false — honesty invariant).
   */
  block_if_applied?: boolean;
}

export interface AntiekBenchRewriteModelDecisionMarketplaceCompose {
  week_id: string;
  week_label: string;
  session_id: string;
  parent_asset_id: string;
  asset_id: string;
  title: string;
  account_id: string;
  focus_task: string;
  rewrite: BenchRewriteProposal;
  model_decision_pack: ModelDecisionTwinSearchHtmlNativeMarketplaceCompose;
  proposal_count: number;
  pack_ready: boolean;
  /** Always false — proposal only; production bench never rewritten. */
  suite_rewritten: false;
  /** Always false — mirrors rewrite.applied. */
  applied: false;
  live_router_authorized: false;
  secrets_stored: false;
  live_meter_read: false;
  remote_index_queried: false;
  backlog_mutated: false;
  store_mutated: false;
  pdf_view_authorized: false;
  pdf_primary: false;
  twin_written: false;
  prompts_injected: false;
  live_dispatch_authorized: false;
  inventory_mutated: false;
  live_dispatched: false;
  pack_dispatched: false;
  merge_executed: false;
  live_execution_authorized: false;
  charge_executed: false;
  draft_written: false;
  record_persisted: false;
  analysis_written: false;
  purchase_executed: false;
  hosted: false;
  remote_fetched: false;
  production_router_verdict: "REJECT";
  notes: string[];
  authority: "antiek_bench_rewrite_model_decision_marketplace_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Antiek-bench recursive rewrite residual stacked on model-decision
 * twin-search HTML-native marketplace pack.
 * Never rewrites suite; never live-routes; never PDF-primary; ND REJECT.
 */
export function composeAntiekBenchRewriteModelDecisionMarketplace(
  input: AntiekBenchRewriteModelDecisionMarketplaceInput,
): AntiekBenchRewriteModelDecisionMarketplaceCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (!input.rewrite || typeof input.rewrite !== "object") {
    throw new Error("rewrite must be an object");
  }
  if (
    !input.model_decision_pack ||
    typeof input.model_decision_pack !== "object"
  ) {
    throw new Error("model_decision_pack must be an object");
  }

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }
  const block_if_applied =
    input.block_if_applied === undefined ? true : input.block_if_applied;
  if (typeof block_if_applied !== "boolean") {
    throw new Error("block_if_applied must be boolean when set");
  }

  const notes: string[] = [
    "suite_rewritten=false · applied=false — rewrite proposal only",
    "live_router_authorized=false · secrets_stored=false · live_meter_read=false",
    "remote_index_queried=false · pdf_primary=false",
    "production_router_verdict=REJECT",
  ];

  const rewrite = proposeAntiekBenchRecursiveRewrite({
    week_label: input.rewrite.week_label,
    patterns: input.rewrite.patterns,
  });
  notes.push(...rewrite.notes.map((n) => `[rewrite] ${n}`));

  if (rewrite.applied !== false) {
    throw new Error("invariant: rewrite.applied must be false");
  }

  const model_decision_pack = composeModelDecisionTwinSearchHtmlNativeMarketplace(
    {
      ...input.model_decision_pack,
      operator_ack: input.operator_ack,
    },
  );
  notes.push(
    ...model_decision_pack.notes.map((n) => `[model_decision_pack] ${n}`),
  );

  const week_label = requireNonEmpty(rewrite.week_label, "week_label");
  const week_id = requireNonEmpty(model_decision_pack.week_id, "week_id");
  const session_id = requireNonEmpty(
    model_decision_pack.session_id,
    "session_id",
  );
  const asset_id = requireNonEmpty(model_decision_pack.asset_id, "asset_id");
  const parent_asset_id = requireNonEmpty(
    model_decision_pack.parent_asset_id,
    "parent_asset_id",
  );
  const title = requireNonEmpty(model_decision_pack.title, "title");
  const account_id = requireNonEmpty(
    model_decision_pack.account_id,
    "account_id",
  );
  const focus_task = requireNonEmpty(
    model_decision_pack.focus_task,
    "focus_task",
  );

  const proposal_count = rewrite.proposals.length;
  const rewrite_ready = proposal_count >= 1;
  const applied_ok = !block_if_applied || rewrite.applied === false;

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      rewrite_ready &&
      applied_ok &&
      model_decision_pack.pack_ready === true &&
      model_decision_pack.suite_rewritten === false &&
      model_decision_pack.live_router_authorized === false &&
      model_decision_pack.secrets_stored === false &&
      model_decision_pack.live_meter_read === false &&
      model_decision_pack.remote_index_queried === false &&
      model_decision_pack.pdf_primary === false &&
      model_decision_pack.pdf_view_authorized === false &&
      model_decision_pack.twin_written === false &&
      model_decision_pack.purchase_executed === false &&
      model_decision_pack.hosted === false &&
      model_decision_pack.inventory_mutated === false &&
      model_decision_pack.production_router_verdict === "REJECT" &&
      input.operator_ack === true;
  } else {
    pack_ready =
      input.operator_ack === true &&
      applied_ok &&
      model_decision_pack.live_router_authorized === false &&
      model_decision_pack.suite_rewritten === false &&
      model_decision_pack.production_router_verdict === "REJECT" &&
      model_decision_pack.pdf_primary === false &&
      (rewrite_ready || model_decision_pack.pack_ready === true);
  }

  if (!rewrite_ready && require_both) {
    notes.push(
      "rewrite has zero proposals — pack_ready blocked (need fail/mixed usage signal)",
    );
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — Antiek-bench rewrite residual + model decision marketplace ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — rewrite, model_decision_pack, or operator_ack gate open",
    );
  }

  if (
    rewrite.applied !== false ||
    model_decision_pack.suite_rewritten !== false ||
    model_decision_pack.live_router_authorized !== false ||
    model_decision_pack.secrets_stored !== false ||
    model_decision_pack.live_meter_read !== false ||
    model_decision_pack.remote_index_queried !== false ||
    model_decision_pack.pdf_primary !== false ||
    model_decision_pack.pdf_view_authorized !== false ||
    model_decision_pack.twin_written !== false ||
    model_decision_pack.purchase_executed !== false ||
    model_decision_pack.hosted !== false ||
    model_decision_pack.inventory_mutated !== false ||
    model_decision_pack.production_router_verdict !== "REJECT"
  ) {
    throw new Error("invariant: honesty flags must remain false / REJECT");
  }

  notes.push("suite_rewritten=false");
  notes.push("applied=false");
  notes.push("live_router_authorized=false");
  notes.push("secrets_stored=false");
  notes.push("live_meter_read=false");
  notes.push("remote_index_queried=false");
  notes.push("backlog_mutated=false");
  notes.push("store_mutated=false");
  notes.push("pdf_view_authorized=false");
  notes.push("pdf_primary=false");
  notes.push("twin_written=false");
  notes.push("prompts_injected=false");
  notes.push("live_dispatch_authorized=false");
  notes.push("inventory_mutated=false");
  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("merge_executed=false");
  notes.push("live_execution_authorized=false");
  notes.push("charge_executed=false");
  notes.push("draft_written=false");
  notes.push("record_persisted=false");
  notes.push("analysis_written=false");
  notes.push("purchase_executed=false");
  notes.push("hosted=false");
  notes.push("remote_fetched=false");
  notes.push("production_router_verdict=REJECT");

  return {
    week_id,
    week_label,
    session_id,
    parent_asset_id,
    asset_id,
    title,
    account_id,
    focus_task,
    rewrite,
    model_decision_pack,
    proposal_count,
    pack_ready,
    suite_rewritten: false,
    applied: false,
    live_router_authorized: false,
    secrets_stored: false,
    live_meter_read: false,
    remote_index_queried: false,
    backlog_mutated: false,
    store_mutated: false,
    pdf_view_authorized: false,
    pdf_primary: false,
    twin_written: false,
    prompts_injected: false,
    live_dispatch_authorized: false,
    inventory_mutated: false,
    live_dispatched: false,
    pack_dispatched: false,
    merge_executed: false,
    live_execution_authorized: false,
    charge_executed: false,
    draft_written: false,
    record_persisted: false,
    analysis_written: false,
    purchase_executed: false,
    hosted: false,
    remote_fetched: false,
    production_router_verdict: "REJECT",
    notes,
    authority:
      "antiek_bench_rewrite_model_decision_marketplace_compose_advisory",
  };
}

export function formatAntiekBenchRewriteModelDecisionMarketplaceSummary(
  c: AntiekBenchRewriteModelDecisionMarketplaceCompose,
): string {
  const budget =
    c.model_decision_pack.decision.would_exceed === null
      ? "would_exceed=null"
      : `would_exceed=${c.model_decision_pack.decision.would_exceed}`;
  return (
    `pack_ready=${c.pack_ready} · ` +
    `proposals=${c.proposal_count} · applied=false · suite_rewritten=false · ` +
    `decision_ready=${c.model_decision_pack.decision.decision_ready} · ` +
    `model=${c.model_decision_pack.decision.driver.decision.selected_model_id} · ` +
    `${budget} · ` +
    `hits=${c.model_decision_pack.twin_search_pack.hit_count} · ` +
    `week=${c.week_label} · task=${c.focus_task} · ` +
    `verdict=${c.production_router_verdict} · ` +
    `live_router_authorized=false · secrets_stored=false · remote_index_queried=false`
  );
}

File: substrate/antiek_bench_rewrite_model_decision_marketplace_compose.py
"""Antiek-bench recursive rewrite residual over model decision + twin search
HTML-native recursive twin marketplace (pure).

suite_rewritten / applied always False (proposal only).
live_router_authorized / secrets_stored / live_meter_read always False.
remote_index_queried / pdf_primary always False.
production_router_verdict always REJECT.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.antiek_bench_recursive_rewrite import (
    AntiekBenchRewriteError,
    BenchRewriteProposal,
    propose_antiek_bench_recursive_rewrite,
)
from substrate.model_decision_twin_search_html_native_marketplace_compose import (
    ModelDecisionTwinSearchHtmlNativeMarketplaceCompose,
    ModelDecisionTwinSearchHtmlNativeMarketplaceComposeError,
    compose_model_decision_twin_search_html_native_marketplace,
)


class AntiekBenchRewriteModelDecisionMarketplaceComposeError(ValueError):
    """Fail-closed validation for bench rewrite + model decision marketplace pack."""


@dataclass(frozen=True)
class AntiekBenchRewriteModelDecisionMarketplaceCompose:
    week_id: str
    week_label: str
    session_id: str
    parent_asset_id: str
    asset_id: str
    title: str
    account_id: str
    focus_task: str
    rewrite: BenchRewriteProposal
    model_decision_pack: ModelDecisionTwinSearchHtmlNativeMarketplaceCompose
    proposal_count: int
    pack_ready: bool
    suite_rewritten: bool
    applied: bool
    live_router_authorized: bool
    secrets_stored: bool
    live_meter_read: bool
    remote_index_queried: bool
    backlog_mutated: bool
    store_mutated: bool
    pdf_view_authorized: bool
    pdf_primary: bool
    twin_written: bool
    prompts_injected: bool
    live_dispatch_authorized: bool
    inventory_mutated: bool
    live_dispatched: bool
    pack_dispatched: bool
    merge_executed: bool
    live_execution_authorized: bool
    charge_executed: bool
    draft_written: bool
    record_persisted: bool
    analysis_written: bool
    purchase_executed: bool
    hosted: bool
    remote_fetched: bool
    production_router_verdict: str
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "week_id": self.week_id,
            "week_label": self.week_label,
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "asset_id": self.asset_id,
            "title": self.title,
            "account_id": self.account_id,
            "focus_task": self.focus_task,
            "rewrite": self.rewrite.to_dict(),
            "model_decision_pack": self.model_decision_pack.to_dict(),
            "proposal_count": self.proposal_count,
            "pack_ready": self.pack_ready,
            "suite_rewritten": False,
            "applied": False,
            "live_router_authorized": False,
            "secrets_stored": False,
            "live_meter_read": False,
            "remote_index_queried": False,
            "backlog_mutated": False,
            "store_mutated": False,
            "pdf_view_authorized": False,
            "pdf_primary": False,
            "twin_written": False,
            "prompts_injected": False,
            "live_dispatch_authorized": False,
            "inventory_mutated": False,
            "live_dispatched": False,
            "pack_dispatched": False,
            "merge_executed": False,
            "live_execution_authorized": False,
            "charge_executed": False,
            "draft_written": False,
            "record_persisted": False,
            "analysis_written": False,
            "purchase_executed": False,
            "hosted": False,
            "remote_fetched": False,
            "production_router_verdict": "REJECT",
            "notes": list(self.notes),
            "authority": (
                "antiek_bench_rewrite_model_decision_marketplace_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise AntiekBenchRewriteModelDecisionMarketplaceComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_antiek_bench_rewrite_model_decision_marketplace(
    *,
    rewrite: object,
    model_decision_pack: object,
    operator_ack: object,
    require_both: object | None = None,
    block_if_applied: object | None = None,
) -> AntiekBenchRewriteModelDecisionMarketplaceCompose:
    """Bench rewrite residual over model-decision marketplace. Never rewrites suite."""
    if not isinstance(operator_ack, bool):
        raise AntiekBenchRewriteModelDecisionMarketplaceComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(rewrite, dict):
        raise AntiekBenchRewriteModelDecisionMarketplaceComposeError(
            "rewrite must be an object"
        )
    if not isinstance(model_decision_pack, dict):
        raise AntiekBenchRewriteModelDecisionMarketplaceComposeError(
            "model_decision_pack must be an object"
        )

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise AntiekBenchRewriteModelDecisionMarketplaceComposeError(
            "require_both must be boolean when set"
        )
    block_applied = True if block_if_applied is None else block_if_applied
    if not isinstance(block_applied, bool):
        raise AntiekBenchRewriteModelDecisionMarketplaceComposeError(
            "block_if_applied must be boolean when set"
        )

    notes: list[str] = [
        "suite_rewritten=false · applied=false — rewrite proposal only",
        "live_router_authorized=false · secrets_stored=false · live_meter_read=false",
        "remote_index_queried=false · pdf_primary=false",
        "production_router_verdict=REJECT",
    ]

    try:
        rw = propose_antiek_bench_recursive_rewrite(
            week_label=rewrite.get("week_label"),
            patterns=rewrite.get("patterns"),
        )
    except AntiekBenchRewriteError as e:
        raise AntiekBenchRewriteModelDecisionMarketplaceComposeError(str(e)) from e
    notes.extend(f"[rewrite] {n}" for n in rw.notes)

    if rw.applied is not False:
        raise AntiekBenchRewriteModelDecisionMarketplaceComposeError(
            "invariant: rewrite.applied must be false"
        )

    try:
        mdp = compose_model_decision_twin_search_html_native_marketplace(
            decision=model_decision_pack.get("decision"),
            twin_search_pack=model_decision_pack.get("twin_search_pack"),
            operator_ack=operator_ack,
            require_both=model_decision_pack.get("require_both"),
            block_on_budget_exceed=model_decision_pack.get(
                "block_on_budget_exceed"
            ),
        )
    except ModelDecisionTwinSearchHtmlNativeMarketplaceComposeError as e:
        raise AntiekBenchRewriteModelDecisionMarketplaceComposeError(
            str(e)
        ) from e
    notes.extend(f"[model_decision_pack] {n}" for n in mdp.notes)

    week_label = _require_nonempty(rw.week_label, field="week_label")
    week_id = _require_nonempty(mdp.week_id, field="week_id")
    session = _require_nonempty(mdp.session_id, field="session_id")
    asset = _require_nonempty(mdp.asset_id, field="asset_id")
    parent = _require_nonempty(mdp.parent_asset_id, field="parent_asset_id")
    title = _require_nonempty(mdp.title, field="title")
    account = _require_nonempty(mdp.account_id, field="account_id")
    focus = _require_nonempty(mdp.focus_task, field="focus_task")

    proposal_count = len(rw.proposals)
    rewrite_ready = proposal_count >= 1
    applied_ok = (not block_applied) or (rw.applied is False)

    if require:
        pack_ready = (
            rewrite_ready
            and applied_ok
            and mdp.pack_ready is True
            and mdp.suite_rewritten is False
            and mdp.live_router_authorized is False
            and mdp.secrets_stored is False
            and mdp.live_meter_read is False
            and mdp.remote_index_queried is False
            and mdp.pdf_primary is False
            and mdp.pdf_view_authorized is False
            and mdp.twin_written is False
            and mdp.purchase_executed is False
            and mdp.hosted is False
            and mdp.inventory_mutated is False
            and mdp.production_router_verdict == "REJECT"
            and operator_ack is True
        )
    else:
        pack_ready = (
            operator_ack is True
            and applied_ok
            and mdp.live_router_authorized is False
            and mdp.suite_rewritten is False
            and mdp.production_router_verdict == "REJECT"
            and mdp.pdf_primary is False
            and (rewrite_ready or mdp.pack_ready is True)
        )

    if not rewrite_ready and require:
        notes.append(
            "rewrite has zero proposals — pack_ready blocked "
            "(need fail/mixed usage signal)"
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — Antiek-bench rewrite residual + model decision "
            "marketplace ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — rewrite, model_decision_pack, or operator_ack "
            "gate open"
        )

    if (
        rw.applied is not False
        or mdp.suite_rewritten is not False
        or mdp.live_router_authorized is not False
        or mdp.secrets_stored is not False
        or mdp.live_meter_read is not False
        or mdp.remote_index_queried is not False
        or mdp.pdf_primary is not False
        or mdp.pdf_view_authorized is not False
        or mdp.twin_written is not False
        or mdp.purchase_executed is not False
        or mdp.hosted is not False
        or mdp.inventory_mutated is not False
        or mdp.production_router_verdict != "REJECT"
    ):
        raise AntiekBenchRewriteModelDecisionMarketplaceComposeError(
            "invariant: honesty flags must remain false / REJECT"
        )

    notes.extend(
        (
            "suite_rewritten=false",
            "applied=false",
            "live_router_authorized=false",
            "secrets_stored=false",
            "live_meter_read=false",
            "remote_index_queried=false",
            "backlog_mutated=false",
            "store_mutated=false",
            "pdf_view_authorized=false",
            "pdf_primary=false",
            "twin_written=false",
            "prompts_injected=false",
            "live_dispatch_authorized=false",
            "inventory_mutated=false",
            "live_dispatched=false",
            "pack_dispatched=false",
            "merge_executed=false",
            "live_execution_authorized=false",
            "charge_executed=false",
            "draft_written=false",
            "record_persisted=false",
            "analysis_written=false",
            "purchase_executed=false",
            "hosted=false",
            "remote_fetched=false",
            "production_router_verdict=REJECT",
        )
    )

    return AntiekBenchRewriteModelDecisionMarketplaceCompose(
        week_id=week_id,
        week_label=week_label,
        session_id=session,
        parent_asset_id=parent,
        asset_id=asset,
        title=title,
        account_id=account,
        focus_task=focus,
        rewrite=rw,
        model_decision_pack=mdp,
        proposal_count=proposal_count,
        pack_ready=pack_ready,
        suite_rewritten=False,
        applied=False,
        live_router_authorized=False,
        secrets_stored=False,
        live_meter_read=False,
        remote_index_queried=False,
        backlog_mutated=False,
        store_mutated=False,
        pdf_view_authorized=False,
        pdf_primary=False,
        twin_written=False,
        prompts_injected=False,
        live_dispatch_authorized=False,
        inventory_mutated=False,
        live_dispatched=False,
        pack_dispatched=False,
        merge_executed=False,
        live_execution_authorized=False,
        charge_executed=False,
        draft_written=False,
        record_persisted=False,
        analysis_written=False,
        purchase_executed=False,
        hosted=False,
        remote_fetched=False,
        production_router_verdict="REJECT",
        notes=tuple(notes),
        authority=(
            "antiek_bench_rewrite_model_decision_marketplace_compose_advisory"
        ),
    )


def format_antiek_bench_rewrite_model_decision_marketplace_summary(
    c: AntiekBenchRewriteModelDecisionMarketplaceCompose,
) -> str:
    budget = (
        "would_exceed=null"
        if c.model_decision_pack.decision.would_exceed is None
        else f"would_exceed={c.model_decision_pack.decision.would_exceed}"
    )
    return (
        f"pack_ready={c.pack_ready} · "
        f"proposals={c.proposal_count} · applied=false · suite_rewritten=false · "
        f"decision_ready={c.model_decision_pack.decision.decision_ready} · "
        f"model={c.model_decision_pack.decision.driver.decision.selected_model_id} · "
        f"{budget} · "
        f"hits={c.model_decision_pack.twin_search_pack.hit_count} · "
        f"week={c.week_label} · task={c.focus_task} · "
        f"verdict={c.production_router_verdict} · "
        "live_router_authorized=false · secrets_stored=false · "
        "remote_index_queried=false"
    )


__all__ = [
    "AntiekBenchRewriteModelDecisionMarketplaceCompose",
    "AntiekBenchRewriteModelDecisionMarketplaceComposeError",
    "compose_antiek_bench_rewrite_model_decision_marketplace",
    "format_antiek_bench_rewrite_model_decision_marketplace_summary",
]
